File: netapp-manageability-sdk-9.3/src/sample/DataFabric_Manager/WebServices/CXF/basic/HelloDfm.java
/*
 * $Id:$
 *
 * Copyright (c) 2009 NetApp, Inc. All rights reserved.
 * Specifications subject to change without notice.
 *
 */

import java.util.Map;

import javax.xml.ws.BindingProvider;
import javax.xml.ws.soap.SOAPFaultException;
import org.apache.cxf.configuration.jsse.TLSClientParameters;
import org.apache.cxf.frontend.ClientProxy;
import org.apache.cxf.transport.http.HTTPConduit;
import javax.net.ssl.TrustManager;
import java.security.cert.*;
import javax.net.ssl.*;

import com.netapp.management.v1.*;

/**
 * This class will print the version number of the DFM Server
 *
 * This Sample code is supported from DataFabric Manager 4.0 onwards.
 */
public final class HelloDfm {

    /**
    * The default HTTP and HTTPS port numbers for the DFM server.
    */
    private static final int DEFAULT_DFM_HTTP_PORT = 8088;
    private static final int DEFAULT_DFM_HTTPS_PORT = 8488;

    /**
     * Client interface to the DFM server
     */
    private static DfmInterface dfmInterface;

    public static void main(String args[]) throws Exception {

        if (args.length < 3) {
            printUsageAndExit();
        }

        String dfmServer = args[0];
        String dfmUser = args[1];
        String dfmPwd = args[2];
        boolean useHttps = false;
        boolean ignoreCert = false;
        
        if (args.length > 3) {
            if (!args[3].equals("-s")) {
                printUsageAndExit();
            }
            useHttps = true;
            if(args.length > 4) {
                if(args[4].equals("-i")) {
                    ignoreCert = true;
                }
                else {
                    printUsageAndExit();
                }
            }
        }

        dfmInterface = createDfmInterface(dfmServer, dfmUser, dfmPwd, useHttps, ignoreCert);

        dfmAbout();
    }

    /**
     * Creates the client proxy that you can use to invoke DFM APIs.
     */
    private static DfmInterface createDfmInterface(String dfmServer, String dfmUser, String dfmPwd, boolean useHttps, boolean ignoreCert) {
        String protocol = "http";
        int portno = DEFAULT_DFM_HTTP_PORT;
        
        if (useHttps) {
            protocol = "https";
            portno = DEFAULT_DFM_HTTPS_PORT;
        }

        String url = protocol + "://" + dfmServer + ":" + portno + "/apis/soap/v1";

        DfmService ss = new DfmService();
        DfmInterface dfmInterface = ss.getDfmPort();

        BindingProvider provider = (BindingProvider) dfmInterface;
        Map<String, Object> reqContext = provider.getRequestContext();
        reqContext.put(BindingProvider.USERNAME_PROPERTY, dfmUser);
        reqContext.put(BindingProvider.PASSWORD_PROPERTY, dfmPwd);
        reqContext.put(BindingProvider.ENDPOINT_ADDRESS_PROPERTY, url);

        if (useHttps && ignoreCert) {
            HTTPConduit httpConduit = (HTTPConduit) ClientProxy.getClient(dfmInterface).getConduit();
            TLSClientParameters tlsParams = new TLSClientParameters();

            // Create a trust manager that will trust all the certificates supplied by the server.
            TrustManager[] trustAllCerts = new TrustManager[] { 
                    new X509TrustManager() {
                        public X509Certificate[] getAcceptedIssuers() {
                            return null;
                        }
                        public void checkClientTrusted(java.security.cert.X509Certificate[] certs, String authType) {
                        }
                        public void checkServerTrusted(java.security.cert.X509Certificate[] certs, String authType) {
                        }
                    } 
            };
            // Trust the servers host name against that certificate.
            tlsParams.setDisableCNCheck(true);
            tlsParams.setTrustManagers(trustAllCerts);
            httpConduit.setTlsClientParameters(tlsParams); 
        }
        return dfmInterface;
    }

    /**
     * Calls the DfmAbout API and prints the results.
     */
    private static void dfmAbout() {
        try {
            // Creating a dfmAbout instance
            DfmAbout dfmAboutParam = new DfmAbout();
            // Invoking the dfm about API and capturing the output datastructure
            DfmAboutResult dfmAboutRes = dfmInterface.dfmAbout(dfmAboutParam);
            // Extracting and printing the version info from dfmAboutRes output
            // structure returned by the API
            System.out.println("\nHello world!  DFM Server version is: " + dfmAboutRes.getVersion());
            
        } catch (SOAPFaultException se) {
            // printing error string if any. the string has the error code and
            // the error description
            System.err.println(se.getFault().getFaultString());
        }
    }

    private static void printUsageAndExit() {
        System.out.println("Usage: HelloDfm <dfmserver> <user> <passwd> [ -s [ -i ] ]\n");
        System.out.println("  -s          Use HTTPS");
        System.out.println("  -i          Ignore server certificate validation for HTTPS\n");
        System.exit(1);
    }

}

File: netapp-manageability-sdk-9.3/src/sample/DataFabric_Manager/WebServices/DotNet/CSharp/basic/HelloDfm/HelloDfm.cs
//============================================================//
//                                                            //
//                                                            //
// HelloDfm.cs                                                //
//                                                            //
// Copyright 2010 NetApp. All rights reserved.                // 
// Specifications subject to change without notice.           // 
//                                                            //
// This sample code will print the version number of the      //
// DataFabric Manager server.                                 //
//                                                            //
// This Sample code is supported from                         //
// DataFabric Manager 4.0 onwards.                            // 
//                                                            //
//============================================================//

using System;
using System.Web.Services.Protocols;
using System.Net;
using System.Net.Security;
using System.Security.Cryptography.X509Certificates;

namespace HelloDfm
{
    /// <summary>
    /// This class will print the version number of the DFM Server.
    /// </summary>
    class HelloDfm
    {

        /// <summary>
        /// The default HTTP and HTTPS port numbers for the DFM server.
        /// </summary>
        private static readonly int DEFAULT_DFM_HTTP_PORT = 8088;
        private static readonly int DEFAULT_DFM_HTTPS_PORT = 8488;

        /// <summary>
        /// Client interface to the DFM server.
        /// </summary>
        private static DfmService dfmService;

        /// <summary>
        /// This function will print various usage options.
        /// </summary>
        public static void PrintUsageAndExit()
        {
            System.Console.WriteLine("Usage: HelloDfm <dfmserver> <user> <passwd> [ -s [ -i ] ] \n");
            System.Console.WriteLine("  -s          Use HTTPS transport");
            System.Console.WriteLine("  -i          Ignores server certificate validation process for HTTPS\n");
            System.Environment.Exit(1);
        }

        /// <summary>
        /// Creates the client proxy that you can use to invoke DFM APIs.
        /// </summary>
        private static void CreateDfmService(string dfmServer, string dfmUser, string dfmPasswd, bool useHttps, bool ignoreCert)
        {
            String protocol = "http";
            int portno = DEFAULT_DFM_HTTP_PORT;

            if (useHttps)
            {
                protocol = "https";
                portno = DEFAULT_DFM_HTTPS_PORT;
            }

            String url = protocol + "://" + dfmServer + ":" + portno + "/apis/soap/v1";
            dfmService = new DfmService();
            ICredentials credentials = new NetworkCredential(dfmUser, dfmPasswd);
            dfmService.Credentials = credentials;
            dfmService.Url = url;

            if (useHttps && ignoreCert)
            {
                //Console.WriteLine("Ignoring certificate validation..");
                ServicePointManager.ServerCertificateValidationCallback +=
                new RemoteCertificateValidationCallback(TrustAllServerCertificates);
            }
        }

        /// <summary>
        /// Entry point for HelloDfm.
        /// </summary>
        public static void Main(string[] args)
        {
            if (args.Length < 3)
            {
                PrintUsageAndExit();
            }

            String dfmServer = args[0];
            String dfmUser = args[1];
            String dfmPwd = args[2];
            bool useHttps = false;
            bool ignoreCert = false;

            if (args.Length > 3)
            {
                if (!args[3].Equals("-s"))
                {
                    PrintUsageAndExit();
                }
                useHttps = true;
                if (args.Length > 4 && args[4].Equals("-i"))
                {
                    ignoreCert = true;
                }
            }
            CreateDfmService(dfmServer, dfmUser, dfmPwd, useHttps, ignoreCert);
            DfmAbout();
        }
                
        /// <summary>
        /// Calls the DfmAbout API and prints the results.
        /// </summary>
        public static void DfmAbout()
        {
            try
            {
                DfmAbout dfmAbout = new DfmAbout();
                DfmAboutResult result = dfmService.DfmAbout(dfmAbout);
                Console.WriteLine("Hello world! DFM Server version is: " + result.Version);
            }
            catch (SoapException e)
            {
                Console.Error.WriteLine(e.Message);
            }
            catch (Exception e)
            {
                Console.Error.WriteLine(e.Message);
            }
        }

        /// <summary>
        /// Callback for validating a server certificate. This method will accept the certificate 
        /// sent by the server during handshake.
        /// </summary>
        public static bool TrustAllServerCertificates(object sender, X509Certificate certificate, X509Chain chain, SslPolicyErrors sslPolicyErrors)
        {
            return true;
        }
    }
}

